discord/discord_video_source.cc

```diff
@@ -0,0 +1,255 @@
+               DiscordFrameReleaseCB releaseCB,
+               void* userData);
+
+ protected:
+  friend class WTF::ThreadSafeRefCounted<VideoSourceDelegate>;
+
+  ~VideoSourceDelegate() {}
+
+  void DoRenderFrameOnIOThread(scoped_refptr<media::VideoFrame> video_frame,
+                               base::TimeTicks estimated_capture_time);
+#ifdef OS_WINDOWS
+  void DoNativeFrameOnMediaThread(HANDLE texture,
+                                  gfx::Size size,
+                                  int64_t timestamp_us,
+                                  void* releaseCBVoid,
+                                  void* userData);
+#endif
+
+  scoped_refptr<base::SingleThreadTaskRunner> io_task_runner_;
+  scoped_refptr<base::SequencedTaskRunner> media_task_runner_;
+
+  // |frame_callback_| is accessed on the IO thread.
+  VideoCaptureDeliverFrameCB frame_callback_;
+
+  media::GpuVideoAcceleratorFactories* gpu_factories_;
+  std::unique_ptr<gpu::GpuMemoryBufferSupport> gpu_memory_buffer_support_;
+};
+
+MediaStreamDiscordVideoSource::VideoSourceDelegate::VideoSourceDelegate(
+    scoped_refptr<base::SingleThreadTaskRunner> io_task_runner,
+    VideoCaptureDeliverFrameCB new_frame_callback)
+    : io_task_runner_(io_task_runner),
+      frame_callback_(std::move(new_frame_callback)),
+      gpu_factories_(Platform::Current()->GetGpuFactories()),
+      gpu_memory_buffer_support_(new gpu::GpuMemoryBufferSupport()) {
+  if (gpu_factories_) {
+    media_task_runner_ = gpu_factories_->GetTaskRunner();
+  }
+}
+
+void MediaStreamDiscordVideoSource::VideoSourceDelegate::
+    DoRenderFrameOnIOThread(scoped_refptr<media::VideoFrame> video_frame,
+                            base::TimeTicks estimated_capture_time) {
+  DCHECK(io_task_runner_->BelongsToCurrentThread());
+  TRACE_EVENT0("webrtc", "VideoSourceDelegate::DoRenderFrameOnIOThread");
+  frame_callback_.Run(std::move(video_frame), estimated_capture_time);
+}
+
+#ifdef OS_WINDOWS
+void MediaStreamDiscordVideoSource::VideoSourceDelegate::
+    DoNativeFrameOnMediaThread(HANDLE texture,
+                               gfx::Size size,
+                               int64_t timestamp_us,
+                               void* releaseCBVoid,
+                               void* userData) {
+  DiscordFrameReleaseCB releaseCB =
+      reinterpret_cast<DiscordFrameReleaseCB>(releaseCBVoid);
+  gfx::GpuMemoryBufferHandle buffer_handle{};
+  buffer_handle.type = gfx::DXGI_SHARED_HANDLE;
+  buffer_handle.dxgi_handle = base::win::ScopedHandle(texture);
+
+  if (!gpu_factories_) {
+    releaseCB(userData);
+    return;
+  }
+  auto* sii = gpu_factories_->SharedImageInterface();
+  if (!sii) {
+    releaseCB(userData);
+    return;
+  }
+  auto buffer_impl =
+      gpu_memory_buffer_support_->CreateGpuMemoryBufferImplFromHandle(
+          std::move(buffer_handle), size, gfx::BufferFormat::RGBA_8888,
+          gfx::BufferUsage::GPU_READ, base::DoNothing());
+  auto mailbox = sii->CreateSharedImage(
```

Comment:
I know we're not using the shared texture path atm but it's surprising to me that we'd do this bit per-frame. Is it cheap?

Reply:
Not sure about the cost. The shared texture path was very much in the "try to get it to work at all" phase when I paused work on it. Definitely would be better to avoid this setup for textures we've already seen.